forc-plugins/forc-client/build.rs

```diff
@@ -0,0 +1,69 @@
+use std::fs;
+use std::path::PathBuf;
+
+fn minify_json(json: &str) -> String {
+    let mut result = String::with_capacity(json.len());
+    let mut in_string = false;
+    let mut previous_char: Option<char> = None;
+
+    for c in json.chars() {
+        if in_string {
+            result.push(c);
+            if c == '"' && previous_char != Some('\\') {
+                in_string = false;
+            }
+        } else {
+            match c {
+                '"' => {
+                    result.push(c);
+                    in_string = true;
+                }
+                ' ' | '\n' | '\r' | '\t' => continue, // Skip whitespace
+                _ => result.push(c),
+            }
+        }
+        previous_char = Some(c);
+    }
+    result
+}
+
+fn main() {
+    // Path to the JSON file in the root directory next to the `src` folder
+    let json_path = PathBuf::from("proxy_abi/proxy_contract-abi.json");
+    // If proxy_contract-abi.json is changed, re-run this script
+    println!("cargo:rerun-if-changed=proxy_abi/proxy_contract-abi.json");
+    // Path to the Rust source file that contains the `abigen!` macro that
+    // creates a `ProxyContract`.
+    let source_file_path = PathBuf::from("src/util/tx.rs");
+    // Read the contents of the JSON file
+    let json_content =
+        fs::read_to_string(json_path).expect("Unable to read proxy_contract-abi.json");
+
+    // Minify the JSON content
+    let minified_json = minify_json(&json_content);
+
+    // Read the contents of the source file
+    let mut source_code =
+        fs::read_to_string(&source_file_path).expect("Unable to read source file");
+
+    // Prepare the replacement string for the `abigen!` macro
+    let escaped_json = minified_json.replace('\\', "\\\\").replace('"', "\\\"");
+    let new_abigen = format!(
+        "abigen!(Contract(name = \"ProxyContract\", abi = \"{}\",));",
+        escaped_json
+    );
+
+    // Use a regular expression to find and replace the `abigen!` macro
+    let re = regex::Regex::new(r#"abigen!\(Contract\(name = "ProxyContract", abi = ".*?",\)\);"#)
```

Comment:
Since that's a non-greedy capture, it might need to be anchored to the end of the string, otherwise `",\)\);` appearing inside the abi source will end the match early:

    let re = regex::Regex::new(r#"abigen!\(Contract\(name = "ProxyContract", abi = ".*?",\)\);$"#)

Reply:
That is a good catch! Only valid way `",\)\);` can appear in abi is probably as a doc comment as in any other place that would be non-valid abi. And luckly if that happens this will be a compile error, as the abi will not be valid (validity of the abi is checked during comp time right now).

I tried to anchor to the end but failed to detect the abi that way, probably my regex wizardy is not enough for this 😄 Also the version you gave cannot match the abi as well. So I think this is something we can omit for now at is _nearly_ impossible that this will be creating a problem  for us and given that this is a very short term solution,  So I created #6547

Reply:
Sorry, just saw this message now.

Ah nice, if it's impossible to slip through, then this is probably good enough (though let's hope nothing changes in the future where comments are allowable for debugging etc). Though yes let's have #6547 for future thinking, because the first thing I thought when seeing the regex was "SQL injection"-like attacks - and if I'm thinking that I'm sure automated offensive code scanners are looking for that too 🫣